codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2ConnectionHandlerBuilder.java

```diff
@@ -0,0 +1,399 @@
+     */
+    protected long gracefulShutdownTimeoutMillis() {
+        return gracefulShutdownTimeoutMillis;
+    }
+
+    /**
+     * Sets the graceful shutdown timeout of the {@link Http2Connection} in milliseconds.
+     */
+    protected B gracefulShutdownTimeoutMillis(long gracefulShutdownTimeoutMillis) {
+        this.gracefulShutdownTimeoutMillis = gracefulShutdownTimeoutMillis;
+        return self();
+    }
+
+    /**
+     * Returns if {@link #build()} will to create a {@link Http2Connection} in server mode ({@code true})
+     * or client mode ({@code false}).
+     */
+    protected boolean isServer() {
+        return isServer != null ? isServer : true;
+    }
+
+    /**
+     * Sets if {@link #build()} will to create a {@link Http2Connection} in server mode ({@code true})
+     * or client mode ({@code false}).
+     */
+    protected B server(boolean isServer) {
+        enforceConstraint("server", "connection", connection);
+        enforceConstraint("server", "codec", decoder);
+        enforceConstraint("server", "codec", encoder);
+
+        this.isServer = isServer;
+        return self();
+    }
+
+    /**
+     * Returns the {@link Http2Connection} to use.
+     *
+     * @return {@link Http2Connection} if set, or {@code null} if not set.
+     */
+    protected Http2Connection connection() {
+        return connection;
+    }
+
+    /**
+     * Sets the {@link Http2Connection} to use.
+     */
+    protected B connection(Http2Connection connection) {
+        enforceConstraint("connection", "server", isServer);
+        enforceConstraint("connection", "codec", decoder);
+        enforceConstraint("connection", "codec", encoder);
+
+        this.connection = checkNotNull(connection, "connection");
+
+        return self();
+    }
+
+    /**
+     * Returns the {@link Http2ConnectionDecoder} to use.
+     *
+     * @return {@link Http2ConnectionDecoder} if set, or {@code null} if not set.
+     */
+    protected Http2ConnectionDecoder decoder() {
+        return decoder;
+    }
+
+    /**
+     * Returns the {@link Http2ConnectionEncoder} to use.
+     *
+     * @return {@link Http2ConnectionEncoder} if set, or {@code null} if not set.
+     */
+    protected Http2ConnectionEncoder encoder() {
+        return encoder;
+    }
+
+    /**
+     * Sets the {@link Http2ConnectionDecoder} and {@link Http2ConnectionEncoder} to use.
+     */
+    protected B codec(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder) {
+        enforceConstraint("codec", "server", isServer);
+        enforceConstraint("codec", "connection", connection);
+        enforceConstraint("codec", "frameLogger", frameLogger);
+        enforceConstraint("codec", "headerSensitivityDetector", headerSensitivityDetector);
+        enforceConstraint("codec", "encoderEnforceMaxConcurrentStreams", encoderEnforceMaxConcurrentStreams);
+
+        checkNotNull(decoder, "decoder");
```

Comment:
Understood. My personal preference is to only enforce null checks if they would not otherwise be uncovered in the current scope...less duplication and the user will likely have to look at the code anyways to figure out what "NPE: decoder" means anyways.